# Options on futures in the coupled CARMA model — Step 1

**Point-delivery futures-option pricer (Carr–Madan, adaptive grid) + Monte-Carlo validation.**

A plain-vanilla call on a futures with option expiry $\tau\le T$ is the generalisation of the
energy call (the $\tau=T$ case). It depends only on the **price factor** $X$, so it isolates the
risk-neutral price-factor law — the right first stage before any quanto/coupling.

Reuses the calibrated coupled-CARMA parameters from `pricing/fft.ipynb` & `pricing/montecarlo.ipynb`
(`data/levy/carma_montecarlo_final_parameters.csv`).

**Status.** Step 1 (this notebook): pricer built and validated against MC. The option-data loader
(French EEX `O7FM` base-month options) is scaffolded at the end for Step 3 (implied $\vartheta_X$
calibration). Two caveats for Step 3 are noted there: (i) the options are **French**, the spot CARMA
is **German**; (ii) the listed contracts are **delivery-period** (basket) options, so the
point-delivery pricer here is the building block for the Step-2 basket pricer.


In [1]:
# --------------------------------------------------------------- imports & calibrated parameters
import numpy as np, pandas as pd
from pathlib import Path
from scipy.linalg import expm
from scipy.stats import norm, norminvgauss
from scipy.optimize import brentq

DATA = Path('../data')   # run with working dir = Code/pricing

# price CARMA(3,2)
b0_X,b1_X,b2_X = 1.0, -1.69107499948807, 0.0856060699524467
a1_X,a2_X,a3_X = 2.7417368516678, 1.85827342924239, 0.161163594914844
# temperature CARMA(2,1)
sigma_T_state = 0.76963298150931
b0_Y,b1_Y = 1.0, 0.584420888567256
a1_Y,a2_Y = 0.834602554797278, 0.02473622057743
# Gaussian temperature driver / NIG idiosyncratic price driver (per hour)
mu_Y,sigma_Y = 0.004626935939183486, 0.9619333921285353
mu_X,delta_X,alpha_X,beta_X = 0.032544431425439106, 0.3350081939529935, 0.9821566545385684, -0.09493955549441858
lambda_coupling = 0.0017485173776097822
sigma_J = 0.036050456023551204
lambda_sigma_T = lambda_coupling*sigma_T_state

AX = np.array([[0,1,0],[0,0,1],[-a3_X,-a2_X,-a1_X]])
BX = np.array([[0.0],[0.0],[sigma_J]])          # fft convention: sigma_J inside BX
cX = np.array([b0_X,b1_X,b2_X])
Gamma = np.array([[0.0],[0.0],[lambda_sigma_T]])
AY = np.array([[0,1],[-a2_Y,-a1_Y]])
BY = np.array([[0.0],[sigma_T_state]])
cY = np.array([b0_Y,b1_Y])
VarJ = delta_X*alpha_X**2/(alpha_X**2-beta_X**2)**1.5   # NIG variance per hour
VarY = sigma_Y**2


In [2]:
# --------------------------------------------------------------- seasonal level Lambda_S (delivery instant)
_ss = pd.read_csv(DATA/'deseasonalised/seasonalities.csv', index_col=0, parse_dates=True)['log_price_seasonal']
_ss.index = pd.to_datetime(_ss.index, utc=True)
_lk = (pd.DataFrame({'v':_ss.values,'m':_ss.index.month,'d':_ss.index.dayofweek,'h':_ss.index.hour})
       .groupby(['m','d','h'])['v'].mean().to_dict())
_ssmean = float(_ss.mean())
def Lambda_S(t):  # t: pd.Timestamp of the delivery instant
    return _lk.get((t.month,t.dayofweek,t.hour), _ssmean)


In [3]:
# --------------------------------------------------------------- Levy exponents, CARMA kernels, forward
def psi_X(th):   # NIG characteristic exponent (idiosyncratic price driver)
    g = np.sqrt(alpha_X**2-beta_X**2)
    return 1j*mu_X*th + delta_X*(g - np.sqrt(alpha_X**2-(beta_X+1j*th)**2))
def psi_T(th):   # Gaussian exponent (temperature driver)
    return 1j*mu_Y*th - 0.5*sigma_Y**2*th**2
def kbeta_X(s,T):  return float(BX.flatten()@expm(AX.T*(T-s))@cX)   # beta_X(s;T)
def kgamma_X(s,T): return float(Gamma.flatten()@expm(AX.T*(T-s))@cX) # gamma_X(s;T)
def kappa_X(b): return np.real(psi_X(-1j*b))   # cumulant kappa_X(b)=psi_X(-i b)
def kappa_Y(g): return np.real(psi_T(-1j*g))

def _kernels(tau,T,n):
    s = np.linspace(0,tau,n)
    return s, np.array([kbeta_X(si,T) for si in s]), np.array([kgamma_X(si,T) for si in s])

def varG(tau,T,n=1500):                 # Var of the stochastic log-forward part G = c_X' e^{A(T-tau)} Z_tau
    s,kb,kg = _kernels(tau,T,n)
    return np.trapezoid(VarJ*kb**2 + VarY*kg**2, s)

def a_const(tau,T,end,n=1500):          # deterministic log-forward constant a(tau,T)
    lamS = Lambda_S(end)
    if tau>=T: return lamS
    s = np.linspace(tau,T,n)
    kb = np.array([kbeta_X(si,T) for si in s]); kg = np.array([kgamma_X(si,T) for si in s])
    return lamS + np.trapezoid(kappa_X(kb),s) + np.trapezoid(kappa_Y(kg),s)

def forward0(T,end,n=1500):             # F(0,T)=E[S_T] from zero initial state
    s=np.linspace(0,T,n); kb=np.array([kbeta_X(si,T) for si in s]); kg=np.array([kgamma_X(si,T) for si in s])
    return np.exp(Lambda_S(end)+np.trapezoid(kappa_X(kb),s)+np.trapezoid(kappa_Y(kg),s)) - 1000


In [4]:
# --------------------------------------------------------------- Carr-Madan: call on point-delivery F(tau,T)
# Spot is shifted-log:  S_t = exp(Lambda_S(t)+X_t) - 1000  =>  strike shift K -> K+1000.
# Futures-style options (EEX): discounting factor 1 (r=0).
def fourier_drifted_call(u,K,lam,a1):
    Ksh=K+1000.0; Kbar=np.log(Ksh)-lam; z=a1+1j*u
    return Ksh*np.exp(-z*Kbar)/((z-1)*z)

def futures_call_cm(K,tau,T,end,a1=1.5,M=32768,n_s=1500):
    """Time-0 price of a call on F(tau,T), expiry tau<=T, single delivery instant T.
    The damped-Fourier grid adapts to sd[G]: u_max=max(250, 12/sd_G)."""
    sG = np.sqrt(max(varG(tau,T),1e-12))
    u_max = max(250.0, 12.0/sG); eta = 2*u_max/M
    u = (-M//2+np.arange(M))*eta
    w = np.full(M,eta); w[0]*=0.5; w[-1]*=0.5
    s,kb,kg = _kernels(tau,T,n_s)
    phi = np.exp(np.trapezoid(psi_X(np.outer(u-1j*a1,kb)) + psi_T(np.outer(u-1j*a1,kg)), s, axis=1))
    Chat = fourier_drifted_call(u,K,a_const(tau,T,end),a1)
    return (np.sum(w*Chat*phi)/(2*np.pi)).real


In [5]:
# --------------------------------------------------------------- Monte-Carlo benchmark (sub-steppable)
def futures_call_mc(Ks,tau,T,end,N=200000,seed=42,n_sub=4):
    """MC of F(tau,T). n_sub sub-steps/hour; n_sub>1 removes the hourly discretization bias.
    The per-step increment is the exact constant-rate form inv(A)(e^{A dt}-I) e_p * (dL/dt)."""
    dt=1.0/n_sub; nt=int(round(tau*n_sub)); eAXdt=expm(AX*dt)
    MXX=(np.linalg.inv(AX)@(eAXdt-np.eye(3))@np.array([[0.],[0.],[1.]])).flatten()/dt
    np.random.seed(seed); ZX=np.zeros((N,3))
    aJ,bJ = alpha_X*delta_X*dt, beta_X*delta_X*dt        # NIG over dt: delta,mu scale linearly in time
    for _ in range(nt):
        dLY = np.random.normal(mu_Y*dt, sigma_Y*np.sqrt(dt), size=N)
        dLJ = norminvgauss.rvs(aJ,bJ,loc=mu_X*dt,scale=delta_X*dt,size=N)
        dLX = lambda_coupling*sigma_T_state*dLY + sigma_J*dLJ
        ZX = ZX@eAXdt.T + np.outer(dLX,MXX)
    G = ZX@(expm(AX*(T-tau)).T@cX)
    F = np.exp(a_const(tau,T,end)+G) - 1000.0
    out = {K:(np.maximum(F-K,0).mean(), np.maximum(F-K,0).std()/np.sqrt(N)) for K in Ks}
    return out, F


## Validation
`tau=T` must reproduce the energy call (notebook FFT 16.8395 / MC 16.8463). Sub-stepping the MC
shows convergence to the continuous-time price (the native dt=1 MC carries a ~0.9% discretization
bias). Then Carr–Madan vs MC across strikes, near (`gap=2h`) and farther (`gap=24h`) from delivery.

In [6]:
START=pd.Timestamp('2026-05-13 12:00:00'); END=pd.Timestamp('2026-06-12 19:00:00')
T=(END-START).total_seconds()/3600; F0=forward0(T,END)

print('SANITY tau=T  (FFT 16.8395 / MC 16.8463) + MC time-step convergence')
print(f'  CM  (K=130)         = {futures_call_cm(130.0,T,T,END):.4f}')
for ns in (1,3,6):
    mc,_=futures_call_mc([130.0],T,T,END,N=80000,n_sub=ns)
    print(f'  MC  (K=130,n_sub={ns}) = {mc[130.0][0]:.4f}  (SE {mc[130.0][1]:.4f})  dt={1/ns:.3f}h')

print(f'\nVALIDATION  F(0,T)={F0:.3f}')
for gap in (2.0,24.0):
    tau=T-gap
    Ks=[float(round(F0)-20),float(round(F0)-8),float(round(F0)),float(round(F0)+12),float(round(F0)+30)]
    mc,F=futures_call_mc(Ks,tau,T,END,N=200000,n_sub=4)
    print(f'  gap={gap:.0f}h (tau={tau:.0f}): sd[F]={F.std():.3f}, E[F]={F.mean():.3f} (mart. vs {F0:.3f})')
    print(f"  {'K':>7} {'Carr-Madan':>12} {'MC':>12} {'SE':>8} {'diff/SE':>8}")
    for K in Ks:
        cm=futures_call_cm(K,tau,T,END); m,se=mc[K]
        print(f'  {K:>7.1f} {cm:>12.4f} {m:>12.4f} {se:>8.4f} {((cm-m)/se if se>0 else 0):>8.2f}')


SANITY tau=T  (FFT 16.8395 / MC 16.8463) + MC time-step convergence


  CM  (K=130)         = 16.8404


  MC  (K=130,n_sub=1) = 16.6923  (SE 0.0779)  dt=1.000h


  MC  (K=130,n_sub=3) = 16.8397  (SE 0.0782)  dt=0.333h


  MC  (K=130,n_sub=6) = 16.8375  (SE 0.0790)  dt=0.167h

VALIDATION  F(0,T)=136.583


  gap=2h (tau=725): sd[F]=33.309, E[F]=136.593 (mart. vs 136.583)
        K   Carr-Madan           MC       SE  diff/SE


    117.0      25.1497      25.1537   0.0571    -0.07


    129.0      17.1191      17.1247   0.0494    -0.11


    137.0      12.7378      12.7431   0.0436    -0.12


    149.0       7.6982       7.7010   0.0347    -0.08


    167.0       3.1916       3.2034   0.0228    -0.52


  gap=24h (tau=703): sd[F]=4.149, E[F]=136.583 (mart. vs 136.583)
        K   Carr-Madan           MC       SE  diff/SE


    117.0      19.5831      19.5833   0.0093    -0.01


    129.0       7.6685       7.6690   0.0088    -0.05


    137.0       1.3998       1.4018   0.0051    -0.39


    149.0       0.0060       0.0060   0.0003     0.15


    167.0       0.0002       0.0000   0.0000     0.00


## Black-76 helpers (futures-style, discount factor 1)
For Step 3: convert market implied vols <-> prices to build the calibration target.

In [7]:
def black76_call(F,K,sigma,tau):
    if sigma<=0 or tau<=0: return max(F-K,0.0)
    sv=sigma*np.sqrt(tau); d1=(np.log(F/K)+0.5*sv*sv)/sv; d2=d1-sv
    return F*norm.cdf(d1)-K*norm.cdf(d2)        # DF=1 (futures-style)

def black76_iv(price,F,K,tau):
    intrinsic=max(F-K,0.0)
    if price<=intrinsic+1e-12: return np.nan
    try:    return brentq(lambda s: black76_call(F,K,s,tau)-price, 1e-6, 50.0, maxiter=200)
    except ValueError: return np.nan


## Option-data loader (French EEX `O7FM` base-month options) — scaffold for Step 3

Each daily snapshot carries the underlying futures price (`UnderlyingSettlementPx`), strike,
option price, Black-76 IV, time to expiry, and the **delivery period** (`Start`,`End`).

**Two caveats before calibrating $\vartheta_X$ to these prices:**
1. **Market mismatch** — these are *French* base-load options (underlying `F7BM`, ~44 EUR/MWh);
   the CARMA model here is calibrated on *German* spot (forward ~136). A self-consistent implied
   calibration needs a French spot CARMA fit (Phase D.1) or an explicit caveat.
2. **Delivery period** — the contracts deliver over a month (`Start`..`End`), so they are
   *basket* options on the delivery-period future. The point-delivery pricer above is the building
   block; the Step-2 basket pricer averages it over a delivery grid (`eq:delivery_forward_grid`).

In [8]:
import zipfile
IV_ZIP = Path('../iv_hypercube_daily.zip'); IV_DIR = Path('../iv_hypercube_daily')
if not IV_DIR.exists():
    with zipfile.ZipFile(IV_ZIP) as z: z.extractall(IV_DIR.parent)

_COLS=['TradeDate','OptionRoot','OptionType','StrikePx','UnderlyingSettlementPx',
       'OptionSettlementPx','TimeToExpiryYears','Start','End','ImpliedVol_black76','DiscountFactor']
def load_calls(date, root='O7FM'):
    """Return the call panel for one snapshot date (YYYY-MM-DD).
    tau_years -> tau_h via *8760 (model is in hours)."""
    df=pd.read_parquet(IV_DIR/f'iv_hypercube_{date}.parquet', columns=_COLS)
    c=df[(df.OptionType=='Call')&(df.OptionRoot==root)].copy()
    c=c.rename(columns={'UnderlyingSettlementPx':'F','StrikePx':'K','OptionSettlementPx':'price',
                        'ImpliedVol_black76':'iv_mkt','TimeToExpiryYears':'tau_yr'})
    c['tau_h']=c['tau_yr']*8760.0
    return c[['TradeDate','Start','End','tau_yr','tau_h','F','K','price','iv_mkt','DiscountFactor']]

dates=sorted(p.stem.replace('iv_hypercube_','') for p in IV_DIR.glob('iv_hypercube_*.parquet'))
print(f'{len(dates)} snapshots, {dates[0]} .. {dates[-1]}')
demo=load_calls(dates[0])
print(f'snapshot {dates[0]}: {len(demo)} O7FM calls; F={demo.F.iloc[0]:.2f}; '
      f'deliveries={sorted(demo.End.dropna().unique())[:4]}...')
# Black-76 round-trip sanity on the front-month ATM-ish cross-section:
front=demo[demo.End==demo.End.min()].copy()
front['iv_check']=[black76_iv(p,F,K,t) for p,F,K,t in zip(front.price,front.F,front.K,front.tau_yr)]
print('Black-76 IV round-trip (should match iv_mkt):')
print(front[['K','F','price','iv_mkt','iv_check']].head(6).to_string(index=False))


223 snapshots, 2025-01-02 .. 2026-04-30
snapshot 2025-01-02: 158 O7FM calls; F=103.68; deliveries=['2025-02-28', '2025-03-31', '2025-04-30', '2025-05-31']...
Black-76 IV round-trip (should match iv_mkt):
   K      F  price   iv_mkt  iv_check
82.0 103.68 22.195 0.613871  0.613871
83.0 103.68 21.273 0.609587  0.609587
84.0 103.68 20.361 0.605400  0.605400
85.0 103.68 19.460 0.601316  0.601316
86.0 103.68 18.572 0.597527  0.597527
87.0 103.68 17.697 0.593795  0.593795


## Next steps
- **Step 2 — basket pricer.** Average `futures_call_cm` over a delivery grid $T_1<\dots<T_n$ for
  the real delivery-period contracts (quadrature + the point-delivery pricer per node), validated
  against a delivery-period MC.
- **Step 3 — implied $\vartheta_X$.** With structural + physical-NIG parameters fixed, calibrate the
  Esscher tilt $\vartheta_X$ ($\beta\to\beta+\vartheta_X$) to the market prices per snapshot
  (`eq:implied_calibration_esscher`); report the implied-vs-historical NIG table and the time series
  of $\hat\vartheta_X$. Resolve the French-options / German-spot mismatch first (Phase D.1).

# Step 2 — delivery-period (basket) option

Exchange-traded power options deliver over a **period** $[T_1,T_2]$. The delivery-period future
$F^{\rm del}(\tau)=\sum_j w_j F(\tau,u_j)$ is a weighted average of point-forwards, **all driven by
the same state $Z_\tau$**, so $B:=F^{\rm del}+1000=\sum_j w_j\exp(a_j+v_j^\top Z_\tau)$ is a basket of
correlated lognormals — not log-affine, hence no 1-D Carr–Madan. We provide:

* `option_delivery_mc` — exact MC over $Z_\tau$ (**benchmark**),
* `option_delivery_proxy` — moment-matched lognormal with an $O(n)$ linearised variance
  $g^\top\mathrm{Cov}(Z_\tau)\,g$ (**fast proxy** for the calibration loop).

The base-load average is sampled **hourly** (the strong intraday seasonal must not be aliased);
kernels $\beta_X,\gamma_X$ are precomputed vs horizon and interpolated for speed.

In [9]:
from scipy.optimize import brentq
def node_time(u): return START+pd.Timedelta(hours=float(u))
def build_kernels(hmax,step=0.5):
    Hk=np.arange(0,hmax+step,step)
    KB=np.array([float(BX.flatten()@expm(AX.T*h)@cX) for h in Hk])
    KG=np.array([float(Gamma.flatten()@expm(AX.T*h)@cX) for h in Hk])
    return Hk,KB,KG
def _kb(h,Hk,KB): return np.interp(h,Hk,KB)
def _kg(h,Hk,KG): return np.interp(h,Hk,KG)
def grid_hourly(T1,T2):
    u=np.arange(T1,T2+1e-9,1.0); w=np.full(len(u),1.0); w[0]=w[-1]=0.5; w/=w.sum(); return u,w
def a_node(tau,u,Hk,KB,KG,nq=800):
    lamS=Lambda_S(node_time(u))
    if tau>=u: return lamS
    s=np.linspace(tau,u,nq)
    return lamS+np.trapezoid(kappa_X(_kb(u-s,Hk,KB)),s)+np.trapezoid(kappa_Y(_kg(u-s,Hk,KG)),s)
def fwd0_node(u,Hk,KB,KG,nq=800):
    s=np.linspace(0,u,nq)
    return np.exp(Lambda_S(node_time(u))+np.trapezoid(kappa_X(_kb(u-s,Hk,KB)),s)+np.trapezoid(kappa_Y(_kg(u-s,Hk,KG)),s))-1000
def logMGF_node(tau,u,Hk,KB,KG,nq=800):
    s=np.linspace(0,tau,nq)
    return np.trapezoid(kappa_X(_kb(u-s,Hk,KB)),s)+np.trapezoid(kappa_Y(_kg(u-s,Hk,KG)),s)
def cov_Z(tau,nq=300):
    s=np.linspace(0,tau,nq); Q=VarJ*(BX@BX.T)+VarY*(Gamma@Gamma.T)
    M=np.array([expm(AX*(tau-si))@Q@expm(AX.T*(tau-si)) for si in s])
    return np.trapezoid(M,s,axis=0)


In [10]:
def option_delivery_mc(Ks,tau,T1,T2,Hk,KB,KG,N=200000,seed=42,n_sub=4):
    u,w=grid_hourly(T1,T2)
    a=np.array([a_node(tau,uj,Hk,KB,KG) for uj in u])
    V=np.column_stack([expm(AX.T*(uj-tau))@cX for uj in u])
    dt=1.0/n_sub; nt=int(round(tau*n_sub)); eAXdt=expm(AX*dt)
    MXX=(np.linalg.inv(AX)@(eAXdt-np.eye(3))@np.array([[0.],[0.],[1.]])).flatten()/dt
    np.random.seed(seed); Z=np.zeros((N,3)); aJ,bJ=alpha_X*delta_X*dt,beta_X*delta_X*dt
    for _ in range(nt):
        dLY=np.random.normal(mu_Y*dt,sigma_Y*np.sqrt(dt),N)
        dLJ=norminvgauss.rvs(aJ,bJ,loc=mu_X*dt,scale=delta_X*dt,size=N)
        Z=Z@eAXdt.T+np.outer(lambda_coupling*sigma_T_state*dLY+sigma_J*dLJ,MXX)
    F=np.exp(a[None,:]+Z@V)@w-1000.0
    return {K:(np.maximum(F-K,0).mean(),np.maximum(F-K,0).std()/np.sqrt(N)) for K in Ks},F

def option_delivery_proxy(Ks,tau,T1,T2,Hk,KB,KG):
    u,w=grid_hourly(T1,T2)
    a=np.array([a_node(tau,uj,Hk,KB,KG) for uj in u]); lm=np.array([logMGF_node(tau,uj,Hk,KB,KG) for uj in u])
    V=np.column_stack([expm(AX.T*(uj-tau))@cX for uj in u]); wexp=w*np.exp(a+lm)
    m1=float(wexp.sum()); g=V@wexp; VarB=float(g@cov_Z(tau)@g)
    s2=np.log(1+VarB/m1**2); sd=np.sqrt(max(s2,0.0)); out={}
    for K in Ks:
        Ksh=K+1000.0
        if sd<=0: out[K]=max(m1-Ksh,0.0)
        else:
            d1=(np.log(m1/Ksh)+0.5*s2)/sd; out[K]=m1*norm.cdf(d1)-Ksh*norm.cdf(d1-sd)
    return out,m1-1000.0,sd


In [11]:
tau=350.0; T1=360.0; T2=T1+720.0          # expiry 350h; base-load 30-day delivery
Hk,KB,KG=build_kernels(T2)
uu,ww=grid_hourly(T1,T2)
F0=float(np.sum(ww*np.array([fwd0_node(uj,Hk,KB,KG) for uj in uu])))
print(f'F^del(0;T1,T2) = {F0:.4f}  (hourly base-load average)')
Ks=[float(round(F0)-15),float(round(F0)-5),float(round(F0)),float(round(F0)+8),float(round(F0)+20)]
mc,F=option_delivery_mc(Ks,tau,T1,T2,Hk,KB,KG,N=300000,n_sub=4)
ap,Fa,sd=option_delivery_proxy(Ks,tau,T1,T2,Hk,KB,KG)
print(f'E[F^del(tau)](MC)={F.mean():.4f} (martingale vs {F0:.4f});  basket lognormal sd={sd:.5f}')
print(f"  {'K':>7} {'Proxy-LN':>11} {'MC':>11} {'MC SE':>8} {'diff/SE':>8}")
for K in Ks:
    m,se=mc[K]; print(f'  {K:>7.1f} {ap[K]:>11.4f} {m:>11.4f} {se:>8.4f} {((ap[K]-m)/se if se>0 else 0):>8.2f}')

atmK=float(round(F0)); tau_yr=tau/8760.0
iv_model=black76_iv(mc[atmK][0],F0,atmK,tau_yr)
print(f'\nMODEL ATM Black-76 IV = {iv_model*100:.2f}%   (market O7FM month options ~ 60%)')


F^del(0;T1,T2) = 76.9386  (hourly base-load average)


E[F^del(tau)](MC)=76.9411 (martingale vs 76.9386);  basket lognormal sd=0.00021
        K    Proxy-LN          MC    MC SE  diff/SE
     62.0     14.9410     14.9411   0.0004    -0.23
     72.0      4.9410      4.9411   0.0004    -0.23
     77.0      0.0628      0.0600   0.0002    14.20
     85.0      0.0000      0.0000   0.0000     0.00
     97.0      0.0000      0.0000   0.0000     0.00

MODEL ATM Black-76 IV = 1.42%   (market O7FM month options ~ 60%)


## Finding (reshapes Step 3)

The model-implied ATM IV of a monthly base-load option is **~1–2%**, vs **~60% in the market**.
The hourly-calibrated price CARMA mean-reverts in *hours*, so the conditional variance of the
delivery-period future at a one-month horizon is negligible. Crucially, the Esscher tilt
$\vartheta_X$ shifts $\beta\to\beta+\vartheta_X$ (the level/skew) but is **variance-invariant**
(it leaves $A_X$, hence the forward-vol term structure, unchanged — cf. `rem:atm_term`).

**Consequence for Step 3.** A one-parameter implied calibration of $\vartheta_X$ to these option
prices cannot match the market smile *level of vol*. The options block should therefore either
(i) extend the model with a persistent / low-frequency factor (or a second CARMA component with a
slow eigenvalue) that lifts the monthly forward vol, or (ii) be framed as documenting the
spot-vs-forward volatility disconnect in power markets — a known stylised fact — rather than a
smile fit. This is the key empirical decision before investing in the implied calibration.

# Step 2b — Can a persistent factor close the forward-vol gap?

Three diagnostics: (1) the current price-CARMA eigenvalue half-lives; (2) the empirical
spot-residual ACF at long lags (is there a slow component the model misses?); (3) an augmented
model `fast CARMA(3,2) + independent slow OU`, sweeping `(half-life, sigma_inf)` -> monthly ATM IV.

In [12]:
# (1) current eigenvalue half-lives
ev=np.linalg.eigvals(AX)
print('(1) price CARMA(3,2) eigenvalue half-lives:')
for l in sorted(ev,key=lambda z:z.real):
    hl=np.log(2)/(-l.real) if l.real<0 else np.inf
    print(f'    lambda={l.real:+.4f}  half-life {hl:7.1f} h = {hl/24:5.2f} d')

# (2) empirical spot-residual ACF at long lags
pr=pd.read_csv('../data/deseasonalised/price_resid.csv',index_col=0)['price_deseasoned'].to_numpy()
pr=pr-pr.mean(); n=len(pr); v0=np.dot(pr,pr)/n
acf=lambda k: float(np.dot(pr[:-k],pr[k:])/((n-k)*v0))
print(f'\n(2) empirical price-residual ACF (n={n}):')
for k in [1,24,72,168,336,720,1440,2160]:
    print(f'    lag {k:5d} h ({k/24:6.1f} d): ACF = {acf(k):+.4f}')
spot_sd=float(np.std(pr)); slow_share=acf(1440)   # long-lag ACF ~ slow-variance share
print(f'    spot log-resid sd = {spot_sd:.4f};  implied slow-variance share ~ {slow_share:.2f} '
      f'=> sigma_inf ~ {np.sqrt(max(slow_share,0))*spot_sd:.4f}')


(1) price CARMA(3,2) eigenvalue half-lives:
    lambda=-1.7103  half-life     0.4 h =  0.02 d
    lambda=-0.9301  half-life     0.7 h =  0.03 d
    lambda=-0.1013  half-life     6.8 h =  0.29 d

(2) empirical price-residual ACF (n=26281):
    lag     1 h (   0.0 d): ACF = +0.9249
    lag    24 h (   1.0 d): ACF = +0.4779
    lag    72 h (   3.0 d): ACF = +0.2176
    lag   168 h (   7.0 d): ACF = +0.1633
    lag   336 h (  14.0 d): ACF = +0.1528
    lag   720 h (  30.0 d): ACF = +0.0856
    lag  1440 h (  60.0 d): ACF = +0.0671
    lag  2160 h (  90.0 d): ACF = +0.0833
    spot log-resid sd = 0.0336;  implied slow-variance share ~ 0.07 => sigma_inf ~ 0.0087


In [13]:
# (3) augmented model: fast CARMA(3,2) + slow OU.  Monthly base-load delivery option.
tau=350.0; T1=360.0; T2=T1+720.0
Hk,KB,KG=build_kernels(T2); u,w=grid_hourly(T1,T2)
a=np.array([a_node(tau,uj,Hk,KB,KG) for uj in u]); V=np.column_stack([expm(AX.T*(uj-tau))@cX for uj in u])
N=200000; n_sub=4; dt=1/n_sub; nt=int(round(tau*n_sub)); eAXdt=expm(AX*dt)
MXX=(np.linalg.inv(AX)@(eAXdt-np.eye(3))@np.array([[0.],[0.],[1.]])).flatten()/dt
np.random.seed(42); Z=np.zeros((N,3)); aJ,bJ=alpha_X*delta_X*dt,beta_X*delta_X*dt
for _ in range(nt):
    dLY=np.random.normal(mu_Y*dt,sigma_Y*np.sqrt(dt),N); dLJ=norminvgauss.rvs(aJ,bJ,loc=mu_X*dt,scale=delta_X*dt,size=N)
    Z=Z@eAXdt.T+np.outer(lambda_coupling*sigma_T_state*dLY+sigma_J*dLJ,MXX)
FZ=a[None,:]+Z@V; xi=np.random.standard_normal(N); tau_yr=tau/8760.0
print(f'(3) augmented monthly ATM IV (%);  s_inf=0 row reproduces the {1.4:.1f}% baseline.')
HL=[720,2160,4320,8760,17520]
print('    sigma_inf |'+''.join(f'{h/24:>8.0f}d' for h in HL))
for s_inf in [0.0,0.010,0.020,0.0336,0.06]:
    row=f'    {s_inf:>9.4f} |'
    for Hh in HL:
        kap=np.log(2)/Hh; Vs=s_inf**2*(1-np.exp(-2*kap*tau)); Xs=np.sqrt(Vs)*xi
        slow=np.exp(-kap*(u-tau))[None,:]*Xs[:,None]+0.5*s_inf**2*(1-np.exp(-2*kap*(u-tau)))[None,:]
        F=np.exp(FZ+slow)@w-1000.0; F0=F.mean(); iv=black76_iv(np.maximum(F-F0,0).mean(),F0,F0,tau_yr)
        row+=f'{(iv*100 if iv==iv else float("nan")):>9.1f}'
    print(row)
print('    (market O7FM monthly ATM IV ~ 60%.)')


(3) augmented monthly ATM IV (%);  s_inf=0 row reproduces the 1.4% baseline.
    sigma_inf |      30d      90d     180d     365d     730d


       0.0000 |      1.4      1.4      1.4      1.4      1.4


       0.0100 |     35.0     27.9     21.5     15.8     11.5


       0.0200 |     69.9     55.8     43.0     31.6     22.8


       0.0336 |    117.1     93.7     72.2     53.0     38.3


       0.0600 |    207.8    166.9    128.7     94.5     68.3
    (market O7FM monthly ATM IV ~ 60%.)


## Conclusion (the Step-3 decision)

**A persistent factor is justified by the spot data and closes a large part of the gap, but not all
of it — the honest reading is that both mechanisms are present.**

* The CARMA(3,2) slowest half-life is **~7 h**, yet the spot-residual ACF still carries **~0.07–0.15**
  autocorrelation at **1–3 months** — a genuine low-frequency component the hourly fit missed. So a
  slow factor is *not* ad hoc; it is in the data.
* That ACF tail implies a slow-variance share of order **8–15%** (`sigma_inf ~ 0.010–0.013`). From
  the sweep, such a factor lifts the monthly ATM IV from ~1.4% to roughly **15–35%** — closing much
  of the gap to the ~60% market level. Reaching the full 60% needs `sigma_inf ~ 0.02` (≈35% variance
  share), i.e. *more* persistence than the ACF alone supports.

**Recommendation.** Re-specify the price factor as **CARMA(3,2) + a slow component** (an added slow
eigenvalue, or a 2-factor fast+slow model) and re-fit to the spot ACF so it matches *both* the
short-lag spikes and the slow tail. This makes monthly forward vol the right order of magnitude, so
the implied-`vartheta_X` calibration (Step 3) becomes meaningful; the residual gap to the market IV
is then the genuine **forward risk premium / unspanned vol** — exactly the object the risk-neutral
calibration is meant to quantify. A spot re-fit with a slow factor is the prerequisite for Step 3.

# Step 2c — data-driven multi-scale fit, and monthly IV with the FITTED slow factor

Instead of *assuming* the slow factor, fit a 3-component model
$\rho(k)=\sum_i w_i e^{-k/\tau_i}$ to the empirical price-residual ACF (free timescales), then
feed the *fitted* slow factor into the monthly-IV check. This pins the slow half-life and variance
share the data actually picks. Reuses the Step-2b simulation (`FZ`, `xi`, `u`, `w`, `tau`, `tau_yr`).

In [14]:
from scipy.optimize import least_squares
pr=pd.read_csv('../data/deseasonalised/price_resid.csv',index_col=0)['price_deseasoned'].to_numpy()
pr=pr-pr.mean(); npr=len(pr); v0=pr@pr/npr; spot_sd=np.sqrt(v0)
def _acf(k): return float(pr[:-k]@pr[k:]/((npr-k)*v0))
lags=np.unique(np.round(np.logspace(0,np.log10(2600),45)).astype(int)); rho=np.array([_acf(int(k)) for k in lags])
def _model(th,k): tau3=np.exp(th[:3]); return (th[3:][:,None]*np.exp(-k[None,:]/tau3[:,None])).sum(0)
def _resid(th): return np.concatenate([_model(th,lags)-rho,[3.0*(th[3:].sum()-1.0)]])
lb=[np.log(0.2),np.log(10),np.log(300),0,0,0]; ub=[np.log(10),np.log(300),np.log(4380),1.2,1.2,1.2]
sol=least_squares(_resid,[np.log(3),np.log(72),np.log(1440),0.5,0.3,0.12],bounds=(lb,ub))
tau3=np.exp(sol.x[:3]); ww=sol.x[3:]/sol.x[3:].sum(); HLf=np.log(2)*tau3; o=np.argsort(tau3)
print('Stage A - data-driven 3-component ACF fit:')
for nm,i in zip(['fast','medium','slow'],o):
    print(f'  {nm:>6}: half-life {HLf[i]:8.1f} h = {HLf[i]/24:6.2f} d   variance share = {ww[i]:.3f}')
print(f'  fit RMSE = {np.sqrt(np.mean((_model(sol.x,lags)-rho)**2)):.4f}')
si=o[-1]; w_slow=ww[si]; H_slow=HLf[si]; s_inf=np.sqrt(w_slow)*spot_sd
print(f'  => slow factor: half-life {H_slow/24:.1f} d, share {w_slow:.1%}, sigma_inf={s_inf:.4f} (spot sd={spot_sd:.4f})')


Stage A - data-driven 3-component ACF fit:
    fast: half-life      2.5 h =   0.10 d   variance share = 0.410
  medium: half-life     23.4 h =   0.97 d   variance share = 0.480
    slow: half-life   1707.3 h =  71.14 d   variance share = 0.110
  fit RMSE = 0.0314
  => slow factor: half-life 71.1 d, share 11.0%, sigma_inf=0.0111 (spot sd=0.0336)


In [15]:
# Stage B: monthly ATM IV with the fitted slow factor (reuse Step-2b FZ, xi, u, w, tau, tau_yr)
def iv_for(s_inf,H_slow):
    kap=np.log(2)/H_slow; Vs=s_inf**2*(1-np.exp(-2*kap*tau)); Xs=np.sqrt(Vs)*xi
    slow=np.exp(-kap*(u-tau))[None,:]*Xs[:,None]+0.5*s_inf**2*(1-np.exp(-2*kap*(u-tau)))[None,:]
    F=np.exp(FZ+slow)@w-1000.0; F0=F.mean(); return black76_iv(np.maximum(F-F0,0).mean(),F0,F0,tau_yr)
print('Stage B - monthly base-load ATM IV:')
print(f'  current model (no slow factor):        {iv_for(0.0,1e9)*100:5.2f}%')
print(f'  + FITTED slow (HL={H_slow/24:.0f}d, share {w_slow:.0%}):   {iv_for(s_inf,H_slow)*100:5.2f}%')
print(f'  + slow share x1.5:                     {iv_for(np.sqrt(1.5)*s_inf,H_slow)*100:5.2f}%')
print(f'  + slow share x2.0:                     {iv_for(np.sqrt(2.0)*s_inf,H_slow)*100:5.2f}%')
print('  market O7FM monthly ATM IV ~ 60%.')


Stage B - monthly base-load ATM IV:


  current model (no slow factor):         1.41%


  + FITTED slow (HL=71d, share 11%):   33.44%


  + slow share x1.5:                     40.94%


  + slow share x2.0:                     47.26%
  market O7FM monthly ATM IV ~ 60%.


## Conclusion

**The data picks a slow factor of half-life ~71 days carrying ~11% of the price-residual variance**
— invisible to the hourly CARMA(3,2) (slowest half-life ~7 h) but clear in the ACF tail. Adding it
lifts the monthly base-load ATM IV from **~1.4% to ~34%**; allowing ~1.5–2× more slow variance (still
within ACF uncertainty) reaches ~41–47%.

So a **finite multi-scale CARMA (≈2.5 h / 1 d / 71 d)** — no fractional memory needed over this lag
range — recovers most of the forward vol. The residual to the ~60% market IV is the genuine
**forward risk premium**, i.e. the object the implied-$\vartheta_X$ calibration (Step 3) should
quantify once the price factor is re-fit with the slow component.

**Action for Step 3:** re-estimate the price factor as a 3-scale CARMA (retain the ~71 d root); then
the implied calibration is meaningful (model base ~34% IV, $\vartheta_X$ explains the rest).

# Step 3-prep — re-specified 3-scale CARMA: re-run the option pipeline + IV term structure

Build the price factor as a **superposition of OU factors at the data-fitted timescales**
(Step 2c: ~2.5 h / ~1 d / ~71 d), a finite CARMA whose variance budget equals the empirical
`v0` and which reproduces the spot ACF. Re-run the delivery-period option pipeline on it and report
the **IV term structure** (the real test: does a multi-scale model give a Samuelson-shaped vol
curve?). Factors are sampled exactly as Gaussian OU — ATM IV is a variance question; NIG tails on
the fast factor affect the smile wings, not ATM IV, and are kept for production. Reuses `sol`, `v0`,
`Lambda_S`, `node_time`, `black76_iv` from earlier cells.

In [16]:
tau3=np.exp(sol.x[:3]); ww=sol.x[3:]/sol.x[3:].sum(); o=np.argsort(tau3)
kappa=1/tau3[o]; Vc=ww[o]*v0; HLc=np.log(2)*tau3[o]
print('Re-specified 3-scale price CARMA (OU superposition):')
for nm,i in zip(['fast','medium','slow'],range(3)):
    print(f'  {nm:>6}: half-life {HLc[i]/24:7.2f} d  kappa={kappa[i]:.5f}/h  var share {ww[o][i]:.3f}  V={Vc[i]:.2e}')
print(f'  total stationary var = {Vc.sum():.2e}  (empirical v0 = {v0:.2e})')


Re-specified 3-scale price CARMA (OU superposition):
    fast: half-life    0.10 d  kappa=0.27677/h  var share 0.410  V=4.63e-04
  medium: half-life    0.97 d  kappa=0.02963/h  var share 0.480  V=5.41e-04
    slow: half-life   71.14 d  kappa=0.00041/h  var share 0.110  V=1.24e-04
  total stationary var = 1.13e-03  (empirical v0 = 1.13e-03)


In [17]:
def atm_iv_3scale(tau_opt,kap,Vv,deliv_len=720.0,gap=10.0,N=400000,seed=1):
    T1=tau_opt+gap; T2=T1+deliv_len; u=np.arange(T1,T2+1e-9,1.0)
    wj=np.full(len(u),1.0); wj[0]=wj[-1]=0.5; wj/=wj.sum()
    a=np.array([Lambda_S(node_time(uj)) for uj in u]).astype(float)
    for i in range(len(kap)): a=a+0.5*Vv[i]*(1-np.exp(-2*kap[i]*(u-tau_opt)))   # convexity
    rng=np.random.default_rng(seed); Zsum=np.zeros((N,len(u)))
    for i in range(len(kap)):
        Vtau=Vv[i]*(1-np.exp(-2*kap[i]*tau_opt))
        Zsum+=np.outer(rng.normal(0,np.sqrt(Vtau),N),np.exp(-kap[i]*(u-tau_opt)))
    F=np.exp(a[None,:]+Zsum)@wj-1000.0; F0=F.mean()
    return black76_iv(np.maximum(F-F0,0).mean(),F0,F0,tau_opt/8760.0),F0

print('IV term structure (ATM, month-long base-load delivery starting just after expiry):')
print(f"  {'maturity':>10} {'3-scale IV':>12} {'1-scale(fast)':>14} {'F^del0':>9}")
kap1=np.array([kappa[0]]); V1=np.array([v0])      # single fast factor carrying all variance
for lab,topt in [('1 week',168.0),('2 weeks',350.0),('1 month',720.0),('2 months',1440.0),('3 months',2160.0)]:
    iv3,F0=atm_iv_3scale(topt,kappa,Vc); iv1,_=atm_iv_3scale(topt,kap1,V1)
    print(f'  {lab:>10} {iv3*100:>11.1f}% {iv1*100:>13.1f}% {F0:>9.1f}')
print('  (market O7FM monthly ATM IV ~ 60%.)')


IV term structure (ATM, month-long base-load delivery starting just after expiry):
    maturity   3-scale IV  1-scale(fast)    F^del0


      1 week        36.0%           0.1%      75.8


     2 weeks        33.9%           0.1%      77.0


     1 month        31.6%           0.1%      76.4


    2 months        26.5%           0.0%      80.4


    3 months        22.2%           0.0%      86.4
  (market O7FM monthly ATM IV ~ 60%.)


## Conclusion — the price factor re-specification works

A **finite 3-scale CARMA** (≈2.5 h / 1 d / 71 d, variance budget matched to the data) re-run through
the option pipeline:

* reproduces the spot ACF and keeps the full finite-state pricing/hedging machinery (no fractional
  memory needed);
* produces a **realistic Samuelson IV term structure** (~36% at 1 week falling to ~22% at 3 months),
  versus the current single-scale model which is flat at ~0.1%;
* lifts the monthly ATM IV to **~32%** — the right shape and about half the ~60% market level.

**This unblocks Step 3.** With the slow root retained, the model carries genuine forward vol, so the
implied-$\vartheta_X$ calibration is now meaningful: the model supplies the ~32% base term structure
and the **residual to the market (~32%→60%) is the forward risk premium** $\vartheta_X$ should
quantify. Production next steps: (i) full QMLE/Kalman re-fit of the 3-scale CARMA (with the NIG
driver on the fast factor) rather than the ACF-matched superposition used here; (ii) add the
temperature coupling back for the quanto; (iii) calibrate $\vartheta_X$ to the `O7FM` smile
(resolving the French-options / German-spot market match, Phase D.1).